File: go/src/models/payment.go
package models

import (
	"payment-microservice/src/processors"
)

// Payment represents a payment transaction
type Payment struct {
	PaymentID  string
	UserID     string
	Amount     float64
	Currency   string
	Status     string
	CreatedAt  *string
	ProcessedAt *string
}

// NewPayment creates a new payment
func NewPayment(paymentID string, userID string, amount float64, currency string) *Payment {
	if currency == "" {
		currency = "USD"
	}
	return &Payment{
		PaymentID:  paymentID,
		UserID:     userID,
		Amount:     amount,
		Currency:   currency,
		Status:     "pending",
		CreatedAt:  nil,
		ProcessedAt: nil,
	}
}

// Process processes the payment
func (p *Payment) Process() (bool, error) {
	p.Status = "processed"
	now := "now"
	p.ProcessedAt = &now
	return true, nil
}

// Refund refunds the payment
func (p *Payment) Refund() (bool, error) {
	p.Status = "refunded"
	return true, nil
}

// CreditCardPayment represents a credit card payment
type CreditCardPayment struct {
	*Payment
	CardNumber string
	CVV        string
	Expiry     string
}

// NewCreditCardPayment creates a new credit card payment
func NewCreditCardPayment(paymentID string, userID string, amount float64, cardNumber string, cvv string, expiry string) *CreditCardPayment {
	return &CreditCardPayment{
		Payment:    NewPayment(paymentID, userID, amount, "USD"),
		CardNumber: cardNumber,
		CVV:        cvv,
		Expiry:     expiry,
	}
}

// Process processes credit card payment
func (p *CreditCardPayment) Process() (bool, error) {
	// Directly accessing external service - tight coupling
	result, err := processors.CreditCardProcessor(p.CardNumber, p.CVV, p.Expiry, p.Amount)
	if err != nil {
		return false, err
	}
	if result {
		p.Status = "processed"
	}
	return result, nil
}

// PayPalPayment represents a PayPal payment
type PayPalPayment struct {
	*Payment
	PayPalEmail string
}

// NewPayPalPayment creates a new PayPal payment
func NewPayPalPayment(paymentID string, userID string, amount float64, paypalEmail string) *PayPalPayment {
	return &PayPalPayment{
		Payment:    NewPayment(paymentID, userID, amount, "USD"),
		PayPalEmail: paypalEmail,
	}
}

// Process processes PayPal payment
func (p *PayPalPayment) Process() (bool, error) {
	// Directly accessing external service - tight coupling
	result, err := processors.PayPalProcessor(p.PayPalEmail, p.Amount)
	if err != nil {
		return false, err
	}
	if result {
		p.Status = "processed"
	}
	return result, nil
}

// Refund represents a refund transaction
type Refund struct {
	RefundID  string
	PaymentID string
	Amount    float64
	Status    string
}

// NewRefund creates a new refund
func NewRefund(refundID string, paymentID string, amount float64) *Refund {
	return &Refund{
		RefundID:  refundID,
		PaymentID: paymentID,
		Amount:    amount,
		Status:    "pending",
	}
}

// Process processes the refund
func (r *Refund) Process() (bool, error) {
	// No validation, no idempotency check
	r.Status = "processed"
	return true, nil
}

// PaymentProcessor interface - NOT USED (violates Open/Closed Principle)
type PaymentProcessor interface {
	Process() (bool, error)
}

File: python/src/models/payment.py
"""
Payment models
"""

class Payment:
    """Represents a payment transaction"""
    
    def __init__(self, payment_id, user_id, amount, currency="USD"):
        self.payment_id = payment_id
        self.user_id = user_id
        self.amount = amount
        self.currency = currency
        self.status = "pending"
        self.created_at = None
        self.processed_at = None
    
    def process(self):
        """Process the payment"""
        self.status = "processed"
        self.processed_at = "now"
        return True
    
    def refund(self):
        """Refund the payment"""
        self.status = "refunded"
        return True


class CreditCardPayment(Payment):
    """Credit card payment"""
    
    def __init__(self, payment_id, user_id, amount, card_number, cvv, expiry):
        super().__init__(payment_id, user_id, amount)
        self.card_number = card_number
        self.cvv = cvv
        self.expiry = expiry
    
    def process(self):
        """Process credit card payment"""
        from src.payment.processors import credit_card_processor
        result = credit_card_processor(self.card_number, self.cvv, self.expiry, self.amount)
        if result:
            self.status = "processed"
        return result


class PayPalPayment(Payment):
    """PayPal payment"""
    
    def __init__(self, payment_id, user_id, amount, paypal_email):
        super().__init__(payment_id, user_id, amount)
        self.paypal_email = paypal_email
    
    def process(self):
        """Process PayPal payment"""
        from src.payment.processors import paypal_processor
        result = paypal_processor(self.paypal_email, self.amount)
        if result:
            self.status = "processed"
        return result


class Refund:
    """Represents a refund transaction"""
    
    def __init__(self, refund_id, payment_id, amount):
        self.refund_id = refund_id
        self.payment_id = payment_id
        self.amount = amount
        self.status = "pending"
    
    def process(self):
        """Process the refund"""
        self.status = "processed"
        return True

